### PR TITLE
Lock file before copy proxy scripts to reduce race condition IDE-141

### PR DESCRIPTION
**Pull request by mustard-mh**
## Description

Implement two things
- Only when handleUri, we start to copy scripts and configure ssh configFile
- Lock file before copy proxy scripts to reduce race condition

## Related Issue(s)

Fixes [IDE-141](https://linear.app/gitpod/issue/IDE-141)

## How to test

- Make sure it can copy scripts

## Documentation

/hold

### CODE REVIEW

File: src/services/localSSHService.ts [hunk 1]
```diff
@@ -60,26 +63,30 @@ export class LocalSSHService extends Disposable implements ILocalSSHService {
...
-        } catch (e) {
-            this.logService.error(e, 'failed to initialize');
-            failureCode = 'Unknown';
-            if (e?.code) {
-                failureCode = e.code;
-            }
-            if (e?.message) {
-                e.message = `Failed to initialize: ${e.message}`;
+        const gitpodHost = this.hostService.gitpodHost;
+        const lockFile = vscode.Uri.joinPath(this.context.globalStorageUri, `${gitpodHost}`);
```
Comment (akosyakov) [inline src/services/localSSHService.ts:67]: maybe add a preifx `.lock`?

**Review by akosyakov — APPROVED** [review]
I did not try, code looks good.

To be clear we expect that copy happens within 10s right? I wonder can we prevent copying at all anything if there are already up-to-date files? It would allow us to save some time?

File: src/services/localSSHService.ts (hunk 1 above)
Comment (jeanp413) [reply src/services/localSSHService.ts:67]: Also move it inside `local-ssh` folder

File: src/services/localSSHService.ts [hunk 2]
```diff
@@ -60,26 +63,30 @@ export class LocalSSHService extends Disposable implements ILocalSSHService {
...
-            this.isSupportLocalSSH = true;
-        } catch (e) {
-            this.logService.error(e, 'failed to initialize');
-            failureCode = 'Unknown';
-            if (e?.code) {
-                failureCode = e.code;
-            }
-            if (e?.message) {
-                e.message = `Failed to initialize: ${e.message}`;
+        const gitpodHost = this.hostService.gitpodHost;
```
Comment (jeanp413) [inline src/services/localSSHService.ts:66]: Shouldn't this be constant? the files copied are always the same independant of host :thinking:

File: src/services/localSSHService.ts (hunk 1 above)
Comment (jeanp413) [inline src/services/localSSHService.ts:67]: Shouldn't the filename be constant? the copied files are independent of host :thinking:

Comment (mustard-mh) [reply src/services/localSSHService.ts:67]: I'm going to change it to
- Copy scripts => constant filename
- Configure ssh => use `gitpodHost`

**Comment by mustard-mh** [comment]
Tested
- Lib `proper-lock` with local script ✅ 
- Smoke test extension that copy will work ✅ 

image

File: src/services/localSSHService.ts [hunk 3]
```diff
@@ -83,12 +87,17 @@ export class LocalSSHService extends Disposable implements ILocalSSHService {
...
 
     private async configureSettings({ proxyScript, launcher }: { proxyScript: string; launcher: string }) {
+        const gitpodHost = this.hostService.gitpodHost;
         const extIpcPort = Configuration.getLocalSshExtensionIpcPort();
-        const hostConfig = this.getHostSSHConfig(this.hostService.gitpodHost, launcher, proxyScript, extIpcPort);
-        await SSHConfiguration.ensureIncludeGitpodSSHConfig();
-        const gitpodConfig = await SSHConfiguration.loadGitpodSSHConfig();
-        gitpodConfig.addHostConfiguration(hostConfig);
-        await SSHConfiguration.saveGitpodSSHConfig(gitpodConfig);
+        const lockName = crypto.createHash('sha1').update(`${gitpodHost}_${extIpcPort}`).digest('hex').substring(0, 8);
```
Comment (jeanp413) [inline src/services/localSSHService.ts:92]: - No need for multiple locks as we are copying both files at the same time 
- Name should be constant, when updating gitpodHost or extIpcPort the event will fire in all opened window

**Comment by mustard-mh** [comment]
I have added a new failureCode `LockFailed` and its error report if you don't think it's harmful then we can filter it (error report) out in next PR.

Going to merge this

```
2023-06-06 01:31:37.680 [trace] Local error report {"component":"vscode-desktop-extension","errorStack":"Error: Failed to lock file: Lock file is already being held\n    at m.lock (/src/services/localSSHService.ts:150:19)\n    at m.prepareInitialize [as initialize] (/src/services/localSSHService.ts:69:9)\n    at /src/remoteConnector.ts:695:27\n\nError: Lock file is already being held\n    at apply (/node_modules/proper-lockfile/lib/lockfile.js:68:47)\n    at s (/node_modules/graceful-fs/polyfills.js:306:20)\n    at FSReqCallback.oncomplete (node:fs:207:5)","version":"0.0.115","workspaceId":"","instanceId":"","userId":"","properties":{"useLocalAPP":"false","common.os":"darwin","common.nodeArch":"arm64","common.platformversion":"22.5.0","common.extname":"gitpod.gitpod-desktop","common.extversion":"0.0.115","common.vscodemachineid":"2d39d84f661d0f4f53d4a5d1284bdd5a91f977b43be5e2fd5784ac8f2ee70f1a","common.vscodesessionid":"9e42961f-bd8c-404b-8421-bf47e6a746001685986189215","common.vscodeversion":"1.78.2","common.product":"desktop","common.uikind":"desktop","error_name":"Error","error_message":"Failed to initialize: Failed to lock file: Lock file is already being held","debug_workspace":"false"}}
2023-06-06 01:31:37.681 [trace] Local event report vscode_desktop_local_ssh_config {"instanceId":"65e48efb-bdf8-48f5-a1f4-333a7f57823b","workspaceId":"gitpodio-gitpodvscodede-2l2n2tealve","debugWorkspace":false,"userId":"261bc022-6360-4e3b-a814-104ff854354b","gitpodVersion":"main-gha.10379","failureCode":"LockFailed","useLocalAPP":"false","status":"failure","common.os":"darwin","common.nodeArch":"arm64","common.platformversion":"22.5.0","common.extname":"gitpod.gitpod-desktop","common.extversion":"0.0.115","common.vscodemachineid":"2d39d84f661d0f4f53d4a5d1284bdd5a91f977b43be5e2fd5784ac8f2ee70f1a","common.vscodesessionid":"9e42961f-bd8c-404b-8421-bf47e6a746001685986189215","common.vscodeversion":"1.78.2","common.isnewappinstall":false,"common.product":"desktop","common.uikind":"desktop","common.remotename":"none"}
```

**Pull request merged**